### PR TITLE
SNOW-3746508 Snowpark Java SDK: add interval type support for UDFs & Sprocs

### PR DESCRIPTION
**Pull request by sfc-gh-jzhan**
1. What Jira ticket or GitHub issue is this PR addressing? Make sure that there is a ticket or issue accompanying your PR.

   Fixes SNOW-3746508 

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

Adds `DayTimeIntervalType` and `YearMonthIntervalType` as first-class types in the Snowpark Java/Scala SDK, enabling them to be declared as UDF and stored procedure parameter and return types.

**Type system additions:**
- `DayTimeIntervalType` / `YearMonthIntervalType` in both the Scala and Java type APIs
- At the UDF/sproc boundary, `INTERVAL DAY TO SECOND` values are transmitted as Arrow `MonthDayNano` and surfaced to handlers as `java.time.Duration`; `INTERVAL YEAR TO MONTH` values are surfaced as `java.time.Period`

**DataFrame API additions:**
- `lit(Duration)` / `lit(Period)` — construct interval literal columns in DataFrame expressions
- `createDataFrame(data, schema)` with interval schema columns — `Duration`/`Period` values are serialized to Snowflake interval string format and CAST back on read

**Result reading:**
- `Row.getDuration()` / `Row.getPeriod()` for typed interval result extraction (requires JDBC 3.27+; `build.sbt` bumped to `3.27.1`)

**Tests:**
- Unit tests covering type parsing, type attribute properties, and Java↔Scala type bridge roundtrips
- E2E integration tests for `lit(Duration|Period)` and `createDataFrame` with interval schema — Scala (`FunctionSuite`, `DataFrameSuite`) and Java (`JavaFunctionSuite`, `JavaDataFrameSuite`); verified passing without any feature flags
- E2E UDF integration tests (`IntervalUDFSuite`, `JavaUDFSuite`) requiring `ENABLE_INTERVAL_TYPES_IN_UDF`; the flag is set at session level via `ALTER SESSION SET` before each test and unset after, so tests run automatically once the parameter is available in the environment

### CODE REVIEW

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala [hunk 1]
```diff
@@ -22,6 +23,55 @@ object DataTypeMapper {
...
+   * ([[Duration.toHoursPart]] and related APIs are Java 9+).
+   */
+  private[snowpark] def formatDuration(duration: Duration): String = {
+    val negative = duration.isNegative
+    val abs = duration.abs()
+    val days = abs.toDays
+    val remSeconds = abs.minusDays(days).getSeconds
+    val hours = remSeconds / 3600
+    val minutes = (remSeconds % 3600) / 60
+    val seconds = remSeconds % 60
```
Comment (sfc-gh-lshi) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala:40]: This seems fishy, don't we need to discard the portion already converted to minutes?

File: src/test/java/com/snowflake/snowpark_test/JavaFunctionSuite.java [hunk 2]
```diff
@@ -3629,6 +3631,27 @@ public void from_utc_timestamp() {
...
+  public void litIntervalTypes() {
+    // Duration: INTERVAL '...' DAY TO SECOND round-trip
+    checkAnswer(
+        getSession().sql("SELECT 1").select(Functions.lit(Duration.ofDays(5))),
+        new Row[] {Row.create(Duration.ofDays(5))});
+    checkAnswer(
+        getSession()
+            .sql("SELECT 1")
+            .select(
+                Functions.lit(Duration.ofDays(2).plusHours(12).plusMinutes(30).plusSeconds(45))),
```
Comment (sfc-gh-lshi) [inline src/test/java/com/snowflake/snowpark_test/JavaFunctionSuite.java:3644]: Can we have a row with some overflow (e.g. 25 hours, 90 minutes, etc) to comprehensively check the dividing/remainder logic we have for the datetime to SQL string conversion

File: src/test/java/com/snowflake/snowpark_test/JavaUDFSuite.java [hunk 3]
```diff
@@ -212,4 +212,45 @@ public void udfNullValue() {
...
+                  DataTypes.YearMonthIntervalType,
+                  DataTypes.YearMonthIntervalType);
+          checkAnswer(
+              df.select(udf.apply(df.col("m"))),
+              new Row[] {Row.create(java.time.Period.of(2, 2, 0))});
+        });
+  }
+
+  private void withIntervalUdfEnabled(Runnable body) {
+    getSession().sql("alter session set ENABLE_INTERVAL_TYPES_IN_UDF = true").collect();
```
Comment (sfc-gh-lshi) [inline src/test/java/com/snowflake/snowpark_test/JavaUDFSuite.java]: Double checking - is this a publicly settable parameter? In the Snowpark client we just check the value of the param and skip the test accordingly, we'll have to do that if public client can't set this

File: src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala [hunk 4]
```diff
@@ -378,6 +380,10 @@ private[snowpark] class ServerConnection(
                     case LongType => data.getLong(resultIndex)
                     case TimestampType => data.getTimestamp(resultIndex)
                     case ShortType => data.getShort(resultIndex)
+                    case DayTimeIntervalType =>
+                      data.getObject(resultIndex, classOf[java.time.Duration])
```
Comment (sfc-gh-yuwang) [inline src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala:384]: getObject() is used here to make sure it is returned in designated type.
while on line 1175 the raw value is returned directly, can we check if this would causing an issue?

can we also add some test for this scenario?

**Review by sfc-gh-yuwang — COMMENTED** [review]
can we add a e2e test for cases of negative year-month type?

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala [hunk 5]
```diff
@@ -116,6 +176,18 @@ object DataTypeMapper {
     dataType match {
       case _ if value == null => "NULL"
       case StringType => s"""'$value'"""
```
Comment (sfc-gh-yuwang) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala:180]: toSqlWithoutCast() this function is used at line 353 of SqlGenerator.scala, and it require the type to be a integral type so the `case DayTimeIntervalType` and `case YearMonthIntervalType` are actually never triggered?

File: build.sbt [hunk 6]
```diff
@@ -78,7 +78,8 @@ lazy val javadocSettings = inConfig(Javadoc)(Defaults.configSettings) ++ Seq(
   Javadoc / packageDoc / artifactName := ((sv, mod, art) =>
     "" + mod.name + "_" + sv.binary + "-" + mod.revision + "-javadoc.jar"))
 
-val jdbcVersion = "3.24.2"
+// 3.27.0+ required for native INTERVAL → java.time.Duration / Period ResultSet reads
```
Comment (sfc-gh-jzhan) [inline build.sbt:81]: I investigate a bit more about the jdbc version, it seems this version bump is un-avoidable:
The 3.24.2 jar has no `IntervalDayTimeToDurationConverter`or `IntervalYearMonthToPeriodConverter` classes. Both appear for the first time in 3.27.1:

`net/snowflake/client/core/arrow/IntervalDayTimeToDurationConverter.class`
`net/snowflake/client/core/arrow/IntervalYearMonthToPeriodConverter.class`

These are the Arrow → Java type converters that back `getObject(index, Duration.class)` / `getObject(index, Period.class)`in older jdbc driver version.

@sfc-gh-yuwang  You mentioned there might be sproc incompatibility if we bump the jdbc version, is existing CI test good enough to catch any incompatibility?

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala (hunk 5 above)
Comment (sfc-gh-jzhan) [reply src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala:180]: Fixed!

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala (hunk 1 above)
Comment (sfc-gh-jzhan) [reply src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala:40]: Using Duration type built-in function will look more elegant here but it only exist after Java9 (this repo is currently at Java8). So unfortunately we will have to do the math here. 

The math should be correct here(`remSeconds % 60` is identical to `(remSeconds % 3600) % 60` — since 3600 is a multiple of 60, the hours portion contributes nothing under % 60). I have also added overflow test cases (Duration.ofHours(25), Duration.ofMinutes(90)) to verify this explicitly.

File: src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala (hunk 4 above)
Comment (sfc-gh-jzhan) [reply src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala:384]: Fixed! `INTERVAL_DAY_TIME` and `INTERVAL_YEAR_MONTH` in `convertToSnowparkValue` now convert the raw Arrow `Long/Integer` to `Duration/Period`.

File: src/test/java/com/snowflake/snowpark_test/JavaUDFSuite.java (hunk 3 above)
Comment (sfc-gh-jzhan) [reply src/test/java/com/snowflake/snowpark_test/JavaUDFSuite.java]: I will send out a PR in GS to enable the parameter and remove the alter param code block.

File: src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala (hunk 4 above)
Comment (sfc-gh-jzhan) [reply src/main/scala/com/snowflake/snowpark/internal/ServerConnection.scala:384]: Added `test("getAs with structured interval fields")` in `RowSuite.scala`

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala (hunk 1 above)
Comment (sfc-gh-lshi) [reply src/main/scala/com/snowflake/snowpark/internal/analyzer/DataTypeMapper.scala:40]: You're right! Thanks for adding the tests so that's it's validated explicitly

File: build.sbt (hunk 6 above)
Comment (sfc-gh-yuwang) [reply build.sbt:81]: yes, exisiting CI would cover the incompatibility, the sproc test would fail if there is any.
you can rebase to avoid noise in the merge gate

**Review by sfc-gh-yuwang — APPROVED** [review]
